### PR TITLE
[SofaSimulationCore] Simplify repetitive code in BaseMechanicalVisitor

### PR DESCRIPTION
**Pull request by alxbilger**
In this PR:

- Added static class member `Visitor::runVisitorTask`
- Used `Visitor::runVisitorTask` in BaseMechanicalVisitor to replace all the boilerplate code
- Simplify the logic in `BaseMechanicalVisitor::processNodeTopDown` even though I have some concerns
- `Visitor::for_each` is now static
- `Visitor::for_each` accepts an optional string parameter giving the type of visitor "fwd" or "bwd"
- `Visitor::for_each_r` is removed in favour of an overload of `Visitor::for_each`
- SOFA_VERBOSE_TRAVERSAL surrounds the calls of `debug_write_state_before` and `debug_write_state_after`, instead of the functions themselves. Doing that, it allows to actually compile the debug content of both function. And I noticed that they had compilation errors (we could not detect it as SOFA_VERBOSE_TRAVERSAL is never defined).

My concerns:

1. I wanted to use `Visitor::runVisitorTask` in other visitors than BaseMechanicalVisitor. But the logic is not the same. BaseMechanialVisitor is the only visitor checking the tags. Should we always check the tags? Also, the debug boilerplate code is not consistent over all the visitors.
2. `Visitor::for_each` returns only the result of the last call in the loop. In another loop on the solvers at the beginning of `BaseMechanicalVisitor::processNodeTopDown`, we check the intermediate values. It is not consistent. I suspect that `Visitor::for_each` must also check intermediate values.
3. In `BaseMechanicalVisitor::processNodeTopDown`, between the process of `mechanicalMapping` and `mechanicalState`, we don't check if `res == RESULT_PRUNE`. It means we do the process of `mechanicalState` even if the process of `mechanicalMapping` returned `RESULT_PRUNE`. It's not consistent with the rest of the code.
4. It is not possible to define SOFA_VERBOSE_TRAVERSAL through cmake

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

**Comment by alxbilger** [comment]
Unit and scene tests are not happy. Setting back to WIP

**Comment by alxbilger** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.cpp [hunk 1]
```diff
@@ -46,199 +46,123 @@ namespace sofa::simulation
...
-                if(testTags(node->mechanicalState))
+                if (testTags(node->mechanicalState))
                 {
-                    ctime_t t = begin(node, node->mechanicalState, "bwd");
-                    this->bwdMappedMechanicalState(ctx, node->mechanicalState);
-                    end(node, node->mechanicalState, t);
-                    t = begin(node, node->mechanicalMapping, "bwd");
-                    this->bwdMechanicalMapping(ctx, node->mechanicalMapping);
-                    end(node, node->mechanicalMapping, t);
+                    runVisitorTask(this, ctx, &BaseMechanicalVisitor::bwdMappedMechanicalState, &*node->mechanicalState, bwdVisitorType);
```
Comment (fredroy) [inline SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.cpp:146]: Maybe move the first runVisitorTask on mappedMS out of the tag test (as runVisitor will also test the tags)...
but this comment is mostly to show I did a thorough review 😁

File: SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.h [hunk 2]
```diff
@@ -240,6 +240,10 @@ class SOFA_SIMULATION_CORE_API BaseMechanicalVisitor : public Visitor
     sofa::helper::vector< sofa::core::ConstMultiVecId > readVector;
     sofa::helper::vector< sofa::core::MultiVecId > writeVector;
 #endif
+
+private:
+    static const std::string fwdVisitorType;
```
Comment (fredroy) [inline SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.h:245]: it wont change anything I suppose but constexpr may be better in this case ? (and you can also inline initialize the value here)

File: SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/Visitor.h [hunk 3]
```diff
@@ -235,8 +230,79 @@ class SOFA_SIMULATION_CORE_API Visitor
...
+    for (const auto& element : list)
+    {
+        res = runVisitorTask(visitor, ctx, task, &*element, typeInfo);
+    }
+    return res;
+}
+
+} // namespace sofa::simulation
 
 #endif
```
Comment (fredroy) [inline SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/Visitor.h]: if you go for a clean, you may remove the ifdef/endif guard with a pragma

File: SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.cpp (hunk 1 above)
Comment (alxbilger) [reply SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.cpp:146]: In any case, the tag test will be performed twice

File: SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.h (hunk 2 above)
Comment (alxbilger) [reply SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.h:245]: constexpr is not allowed on strings. It is allowed from C++20

File: SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.cpp (hunk 1 above)
Comment (fredroy) [reply SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.cpp:146]: true that

File: SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.h (hunk 2 above)
Comment (fredroy) [reply SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/BaseMechanicalVisitor.h:245]: right, I forgot that.

File: SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/Visitor.h (hunk 3 above)
Comment (alxbilger) [reply SofaKernel/modules/SofaSimulationCore/src/sofa/simulation/Visitor.h]: done

**Review by fredroy — APPROVED** [review]

**Pull request merged**